drivers/soundwire/cadence_master.c

```diff
@@ -1248,19 +1271,28 @@ EXPORT_SYMBOL(cdns_set_sdw_stream);
  * @num: Number of PDIs
  * @pdi: PDI instances
  *
- * Find and return a free PDI for a given PDI array
+ * Find and return a free PDI for a given PDI array. The PDI num and
+ * dai_id are expected to match, flag an error otherwise.
  */
 static struct sdw_cdns_pdi *cdns_find_pdi(struct sdw_cdns *cdns,
+					  unsigned int offset,
 					  unsigned int num,
-					  struct sdw_cdns_pdi *pdi)
+					  struct sdw_cdns_pdi *pdi,
+					  int dai_id)
 {
 	int i;
 
-	for (i = 0; i < num; i++) {
-		if (pdi[i].assigned)
-			continue;
-		pdi[i].assigned = true;
-		return &pdi[i];
+	for (i = offset; i < offset + num; i++) {
+		if (pdi[i].num == dai_id) {
+			if (!pdi[i].assigned) {
+				pdi[i].assigned = true;
+				return &pdi[i];
+			}
+
+			dev_err(cdns->dev,
+				"PDI num %d is already allocated\n",
+				dai_id);
```

Comment:
@plbossart I think we should return &pdi[i] even the PDI is assigned. And it is not an error if the PDI is already assigned since we will call cdns_find_pdi() in each playback/capture and the PDI will be assigned at the very first time when cdns_find_pdi() is called. Maybe we can remove the assigned flag at all?

Reply:
ok, i'll get rid of it. it's useless indeed.